src/Orleans.Core/Messaging/Message.cs

```diff
@@ -384,138 +335,71 @@ static bool Append(ref Span<char> dst, ReadOnlySpan<char> value)
+        [StructLayout(LayoutKind.Explicit)]
+        internal struct PackedHeaders
         {
-            return new Message
-            {
-                Category = request.Category,
-                Direction = Message.Directions.Response,
-                Result = Message.ResponseTypes.Error,
-                BodyObject = Response.FromException(exception)
-            };
-        }
+            private const uint DirectionMask = 0x000F_0000;
+            private const int DirectionShift = 16;
+            private const uint ResponseTypeMask = 0x00F0_0000;
+            private const int ResponseTypeShift = 20;
 
-        [Flags]
-        public enum Headers
-        {
-            NONE = 0,
-            ALWAYS_INTERLEAVE = 1 << 0,
-            CACHE_INVALIDATION_HEADER = 1 << 1,
-            CATEGORY = 1 << 2,
-            CORRELATION_ID = 1 << 3,
-            DEBUG_CONTEXT = 1 << 4, // No longer used
-            DIRECTION = 1 << 5,
-            TIME_TO_LIVE = 1 << 6,
-            FORWARD_COUNT = 1 << 7,
-            NEW_GRAIN_TYPE = 1 << 8,
-            GENERIC_GRAIN_TYPE = 1 << 9,
-            RESULT = 1 << 10,
-            REJECTION_INFO = 1 << 11,
-            REJECTION_TYPE = 1 << 12,
-            READ_ONLY = 1 << 13,
-            RESEND_COUNT = 1 << 14, // Support removed. Value retained for backwards compatibility.
-            SENDING_ACTIVATION = 1 << 15,
-            SENDING_GRAIN = 1 << 16,
-            SENDING_SILO = 1 << 17,
-            //IS_NEW_PLACEMENT = 1 << 18,
-
-            TARGET_ACTIVATION = 1 << 19,
-            TARGET_GRAIN = 1 << 20,
-            TARGET_SILO = 1 << 21,
-            TARGET_OBSERVER = 1 << 22,
-            IS_UNORDERED = 1 << 23,
-            REQUEST_CONTEXT = 1 << 24,
-            INTERFACE_VERSION = 1 << 26,
-
-            CALL_CHAIN_ID = 1 << 29,
-
-            INTERFACE_TYPE = 1 << 31
-            // Do not add over int.MaxValue of these.
-        }
+            public static implicit operator PackedHeaders(uint fields) => new() { _fields = fields };
+            public static implicit operator uint(PackedHeaders value) => value._fields;
 
-        internal Headers GetHeadersMask()
-        {
-            var headers = Headers.NONE;
-            if (Category != default)
-            {
-                headers |= Headers.CATEGORY;
-            }
+            // 32 bits: HHHH_HHHH RRRR_DDDD FFFF_FFFF FFFF_FFFF
+            // F: 16 bits for MessageFlags
+            // D: 4 bits for Direction
+            // R: 4 bits for ResponseType
+            // H: 8 bits for ForwardCount (hop count)
+            [FieldOffset(0)]
+            private uint _fields;
 
-            headers = _direction == null ? headers & ~Headers.DIRECTION : headers | Headers.DIRECTION;
+            [FieldOffset(0)]
+            private MessageFlags _flags;
 
-            if (IsReadOnly)
-            {
-                headers |= Headers.READ_ONLY;
-            }
+            [FieldOffset(24)]
```

Comment:
Field offset is measured in bytes, not bits. This should be `[FieldOffset(3)]`

Reply:
What a blunder!